e2e_tests/tests/fixtures/metric_maker/metric_maker.py

```diff
@@ -68,17 +68,51 @@ class MetricMaker(det.CallbackTrialController):
     based on hyperparameters.
     """
 
-    def __init__(self, *args: Any, **kwargs: Any) -> None:
-        super().__init__(*args, **kwargs)
+    def __init__(
+        self,
+        prof: profiler.ProfilerAgent,
```

Comment:
I think you have expanded a lot of `*args, **kwargs`, which is a sign that you have broken a pattern that previously worked.  Looking at the ordering of the `prof` argument, it looks inconsistent across various TrialController classes, which is probably the problem.

\<aside>

You are sort of wrestling with an architectural issue in the harness; we've had this problem where we pass way too many parameters through way too many function calls ever since I did the refactor of the harness a year ago.  It's definitely an antipattern, but it's not an antipattern I had an answer for until we decided on push architecture.  I *think* that antipattern is going to be resolved, at least partially, during the harness refactor I'm working on right now for push architecture.

The root of the problem in today's harness is that the part of the system which needs the most context is the lowest layer of the system, the TrialController.  So it ends up needing all of this information passed into it, since all the information is instantiated elsewhere.

A cleaner pattern would be to let the TrialController own more of the training.  After launch layer happens, it will; the horovod launch layer will instantiate the workers, and inside each worker `harness.py` will figure out just enough information to load the user's Trial class, and call TrialController.from_trial().  That will have much fewer args, because a bunch of the information will then be instantiated *inside* the TrialController.  The overall design should be analagous to where the generic API where after the launch layer, the user's workers each call `ctx = det.generic.init()` to instantiate the same collection of information that we currently have.

This will all be much easier with push architecture because all of the logic which currently is embedded in processing a stream of workloads will instead be part of the generic API's calls for reporting metrics and checkpoints.

\</aside>

So I don't think you should reorder the `prof` arg to fix the problem, I think you should just instantiate the profiler in the base `TrialController.__init__()` so all TrialController subclasses can access it as `self.prof`.  Then, (because we separated the `prof.start()` from `prof.__init__()`) we can put the `with prof():` context manager in each trial controller's `run()` call... or actually only  in the three you care about, which are PyTorchTrialController, TFKerasTrialController, and EstimatorTrialController.  You'll have to tweak the logic very slightly, so that e.g. the `local_rank=0` worker collects system metrics for each compute node.

For reference, a similar effort is underway to move the storage manager into the TrialController.  With the storage manager there, the WorkloadManager can go away entirely, and we will have an easy transition to using the new push architecture endpoints.

If you are curious, I'd be happy to explain more about the upcoming architecture changes via zoom.

Reply:
(at the design, still getting it working with tests and stuff)